### ISSUE TITLE
Strings with Space/Whitespace-Padding won't be parsed correctly

### ISSUE THREAD

**Issue by marcoemrich** [patrick-steele-idem/morphdom#51]
This works:

```
morphdom($("#out"),
  `<div id="out">
    Foo
  </div>`
```

Howerever this doesn't:

```
morphdom($("#out"), `
  <div id="out">
    Foo
  </div>`

```
(not the position of the opening Backtick-Quote)

This als won't work:

```
morphdom($("#out"), ` <div id="out">Foo</div>`
```

(Note the leading Space in the string)

**Comment by patrick-steele-idem** [comment]
Hey @marcoemrich, it makes sense that `morphdom` would not work since you are trying to morph a single DOM element to text (whitespace) + a DOM element. We could update `morphdom` to trim the target string but I feel like that is not the right thing to do since I don't think there should be extra whitespace passed to morphdom in practice. Out of curiosity, why do you have extra whitespace?

**Comment by marcoemrich** [comment]
It's mostly a matter of style. I'm used to write

```
morphdom($("#out"), `
         <div id="out">
            <table id="temperature_table">
              ...
            </table>
          </div>
`);

```

Instead of

```
morphdom($("#out"), 
         `<div id="out">
            <table id="temperature_table">
              ...
            </table>
          </div>
`);

```

The reason is to have a clearer html structure, that is not burdened by the quote-Symbol (looking similar to jsx). Personally I don't a have big problem to switch to the second style.

I still think, however, that it might be a problem, because it violates the "principle of least surprise". If some dev just naively uses the same style I do, he might trip over the issue. The code just doesn't work and it isn't obvious why (the js-error in morphdom doesn't help).

I personally would do the trimming, so a beginner does not have to worry. On the other hand, your explanation regarding the DOM makes sense. I would at very least add a custom error message including something like "consider removing the whitespace at the beginning of string foo". This would make morphdom more beginner friendly.

**Comment by patrick-steele-idem** [comment]
You are right that the currently the code does fail in a surprising way. The current code is blindly picking the first node in the parsed target DOM: https://github.com/patrick-steele-idem/morphdom/blob/0814736664841bddf256bd23d17e655ce36a6d02/lib/index.js#L41

We probably should do something similar to the following:

```javascript
function toElement(str) {
    if (!range && document.createRange) {
        range = document.createRange();
        range.selectNode(document.body);
    }

    var fragment;
    if (range && range.createContextualFragment) {
        fragment = range.createContextualFragment(str);
    } else {
        fragment = document.createElement('body');
        fragment.innerHTML = str;
    }

    var childNodes = fragment.childNodes;
    if (childNodes.length > 1) {
        throw new Error('Target HTML contains multiple root nodes');
    }
    return childNodes[0];
}
```

We would you be interested in trying out that change (or a similar change) and sending a PR if it looks good?

**Comment by marcoemrich** [comment]
sure. If I find the time, I will take a look tomorrow.

**Comment by AutoSponge** [comment]
I also ran into this when I'm mounting a server-rendered div with whitespace in it.

To solve this problem, I wrote my render function with a bit of code like this:
```javascript
if (child.nodeType !== elm.nodeType) {
  container.replaceChild(elm, child)
} else {
  morphdom(child, elm      
}
```

As far as the different lengths issue, the top of my renderer does this:
```javascript
export default (container, ...elms) => {
  const children = [...container.childNodes]
  const length = Math.max(children.length, elms.length)
```

By iterating the `length`, I can decide to append, remove, or morph each elm/child "pair".

**Comment by AutoSponge** [comment]
I've published my render function here: https://github.com/AutoSponge/morphdom-render

You should wrap it with setTimeout or requestAnimationFrame on the browser for best performance.

**Comment by spattersongt** [comment]
Had this issue today.  Willing to test out any patches.  For my issue simply trimming the template string worked

**Comment by AutoSponge** [comment]
`morphdom` probably can't accurately guess which text nodes are meaningful. While I haven't tested it myself, I think the best solution may be to use a minifier/compressor on the HTML/template (see https://github.com/kangax/html-minifier or https://github.com/Swaagie/minimize) and use the same generation methods consistently.
